### PR TITLE
Make DMA helpers device-aware for tensor parallelism

### PR DESCRIPTION
**Pull request by ani300**
## Why this is needed

Tensor-parallel model loading cannot use the normal whole-model `load_model_to_spyre` path. Each rank first shards weights on CPU and then places its shard directly on a specific Spyre device. A plain `.to("spyre")` at that boundary uses the default layout, which is incorrect for access-pattern-specific weights: embeddings need the indirect-access layout used by gather, while linear weights need the matmul-friendly swapped dimension order.

The layout-aware allocator also had a device-semantics bug. Inductor reads the device from every Buffer and passes it to its normal allocation path, but the `FixedTiledLayout` override discarded that device. `spyre_empty_with_layout` had no device parameter and selected the ambient current device. Its `DeviceGuard` was created only after `SpyreStorageImpl`, even though `SpyreAllocator::allocate` reads the guarded current device while constructing that storage. Device selection belongs in the allocator so model-loader DMA calls, generated Inductor buffers, and frontend HBM pools all behave consistently.

Compiling the Gemma 4 MoE and its all-reduce exposed a separate allocation-lifetime bug. `for_each_tile` aliases its rewritten output to the initializer/accumulator allocation. When that accumulator is passed to the Python collective fallback, HBM pool planning excluded only the fallback input name. A differently named alias could still be pooled, relocating the shared allocation. Pool-backed intermediates do not receive standalone Python tensor allocations, so the generated wrapper then called `allreduce_run` with an undefined accumulator such as `buf55`.

## What changed

- add an explicit optional `device` argument to the existing default, linear, embedding, and MoE DMA helpers; no new public wrapper API is introduced
- add an explicit optional device argument to `spyre_empty_with_layout` and install the C++ `DeviceGuard` before storage allocation
- make `FixedTiledLayout` wrapper allocations and frontend HBM-pool allocations pass their real graph device
- normalize integer, string, and `torch.device` destinations and pass the destination through directly
- preserve the requested device in `Tensor.to(..., device_layout=...)`
- require both layout and device equality before taking the no-copy D2D shortcut; cross-device copies return their allocated destination tensor
- implement `SpyreGuardImpl::uncheckedSetDevice` so `c10::DeviceGuard` restores the caller device as required by its contract
- preserve gather-optimized embedding and matmul-optimized linear layouts during direct TP shard placement
- exclude the complete allocation identity of every fallback input from HBM pooling, covering differently named aliases
- add regression coverage for non-current-device allocation, every device-aware DMA layout, same-layout cross-device round trips, generated device arguments, and aliased `ExternKernel` inputs

This lets HF-adapters transfer each CPU TP shard directly into its final on-device layout, while compiled MoE collectives retain a materialized accumulator without adding an eager boundary.

Consumer: https://github.com/torch-spyre/hf-adapters/pull/522

## Testing

- targeted Torch-Spyre tests: 162 passed, 1 skipped, 6 subtests passed
- targeted pre-commit hooks passed, including Ruff and mypy
- HF-adapters tensor-parallel CPU tests: 8 passed
- Gemma 4 26B-A4B TP=2 smoke: passed on both ranks; second pass ~1.41-1.42 s TTFT and ~249-251 ms/token
- Gemma 4 TP=4 smoke passed during the HF-adapters integration

### CODE REVIEW

**Review by dgrove-oss — DISMISSED** [review]
Summary

Two independent commits bundled together. Commit 1 adds public, device-aware DMA entry points (dma_tensor_to_spyre, dma_linear_weight_to_spyre, dma_embedding_weight_to_spyre) so a tensor-parallel loader can place a CPU weight shard directly into its correct on-device layout (matmul-friendly for linear weights, gather-optimal for embeddings) on an explicit destination device, rather than relying on .to("spyre")'s default layout and current-device. Commit 2 fixes an unrelated HBM-pool-planning bug surfaced while compiling Gemma 4 MoE + all-reduce: _is_intermediate excluded fallback-input buffers from pooling only by their exact allocation identity as seen through io_alloc_ids, missing the case where a differently-named alias (produced by for_each_tile's output/accumulator aliasing) shares the same allocation dict — that alias could still get pooled, silently relocating the fallback's actual input storage and producing a reference to a never-materialized tensor (e.g. allreduce_run(buf55) where buf55 doesn't exist).

Review

I traced both fixes against their exact failure mechanisms rather than taking the description at face value:
- Confirmed _target_spyre_device's save/restore of torch.spyre.current_device()/set_device() follows the same global-current-device idiom used elsewhere in this repo (torch.cuda.device()-style), and that current_device/set_device are process-global state, not thread-local — consistent with existing usage, not a new race.
- Verified dma_linear_weight_to_spyre's -> torch.Tensor and dma_embedding_weight_to_spyre's -> torch.Tensor | None annotations each correctly mirror their underlying _dma_to_spyre_dim_order_swapped (never returns None) and _dma_to_spyre_indirect_access (returns None on a non-stick-aligned hidden dim) implementations.
- Confirmed on main that _is_intermediate previously checked only io_alloc_ids, with no reference to fallback reads at all — validating the described gap is real, not already covered elsewhere.
- Confirmed the new fallback_read_alloc_ids check is complementary to, not redundant with, the pre-existing name-based fallback_read subtraction at the all_candidates call site — one catches the exact fallback-read name, the other catches an aliased name sharing its allocation dict.
- Verified _alloc_id's None results are correctly filtered out of fallback_read_alloc_ids via the walrus-operator comprehension, so no spurious None-matches leak into the exclusion set.
- Hand-verified the new test_alias_of_fallback_input_is_not_pool_eligible test actually reproduces the aliasing bug mechanism (shared allocation dict via _make_ftl_buffer_aliased, fallback reading the non-aliased name) rather than a name-only scenario the old code already handled.

BLOCKERs

None found.

SUGGESTIONs

- The new test_public_dma_helpers_accept_an_explicit_device test passes torch.spyre.current_device() as the explicit device, so it verifies the device= parameter is accepted and plumbed through but never actually exercises the set_device/restore branch inside _target_spyre_device (only the device is None and target_index == previous paths run). This mirrors the repo's existing multi-device test conventions (e.g. test_spyre_device_count_and_set_device's min(2, count) guard), so it's not a new gap this PR introduces, but the save/restore logic is exactly the kind of code that's easy to get subtly wrong (e.g. an exception between set_device(target) and the finally), and it's currently untested even on multi-card CI. Worth a follow-up if multi-card CI capacity exists.

QUESTIONs

None — the PR body already explains both fixes' motivation clearly, and the consumer (hf-adapters PR #522) is linked for cross-reference.

Test Coverage

Good for both commits. Commit 1 adds a focused test confirming the public wrappers accept an explicit device, land the tensor there, and preserve the correct specialized device layout for both linear and embedding paths. Commit 2 adds one precise regression test isolating the exact aliasing mechanism the bug report describes, using the existing _make_ftl_buffer_aliased helper and a MagicMock(spec=ExternKernelSchedulerNode) fallback reader — reusing established test infrastructure rather than inventing new scaffolding.

CI / DCO

Both commits signed off. CI just started (20 passing, 0 failing, rest pending at review time) — should be confirmed fully green before merge. Auto-merge is disabled, so this presumably awaits explicit maintainer approval regardless.

Verdict

APPROVE (pending full CI completion, which is in progress with no failures so far).

**Review by thoangtrvn — COMMENTED** [review]
The PR addresses three legitimate bugs:
1. Device-semantics bug in `spyre_empty_with_layout` (device selected after storage allocation)
2. `uncheckedSetDevice` was a no-op, violating `c10::DeviceGuard` contract
3. HBM pool planning alias bug with for_each_tile accumulators

Core fixes are correct and necessary.

---

## Potentially Unnecessary Changes

### 1. New public DMA functions may be over-engineered

The PR adds three new public functions:
- `dma_tensor_to_spyre()`
- `dma_linear_weight_to_spyre()`
- `dma_embedding_weight_to_spyre()`

**Alternative:** The existing underscore-prefixed functions (`_dma_to_spyre_default`, `_dma_to_spyre_dim_order_swapped`, `_dma_to_spyre_indirect_access`) could have been modified in-place to accept the `device` parameter.

**Question:** Do TP users specifically need these public API names, or could they call the private functions directly? My guess is not. 

### 2. `dma_tensor_to_spyre()` appears redundant

```python
def dma_tensor_to_spyre(...):
    """Copy a CPU tensor to a Spyre device using the default device layout."""
    device = _normalize_spyre_device(device)
    return _dma_to_spyre_default(cpu_tensor, target_dtype=target_dtype, device=device)
```

If `_dma_to_spyre_default` now has the `device` parameter, this wrapper provides no additional value. Users could call the private function directly or use `.to("spyre")`.

---

## Cosmetic Changes Unrelated to PR Purpose

### 3. Import reordering in `model_utils.py`

```diff
-import torch.nn as nn
+from torch import nn
```

And reordering imports between `torch` and `torch_spyre._C` blocks. This is cosmetic formatting unrelated to the device-semantics fix.

### 4. Import style change in `test_model_utils.py`

```diff
-import torch.nn as nn
+from torch import nn
```

Same issue — cosmetic import style change.

### 5. Lambda formatting in `_transfer_module`

```diff
-            lambda t: _dma_to_spyre_default(t, target_dtype=dtype)
-            if t is not None and t.device.type != DEVICE_NAME
-            else t
+            lambda t: (
+                _dma_to_spyre_default(t, target_dtype=dtype)
+                if t is not None and t.device.type != DEVICE_NAME
+                else t
+            )
```

Formatting-only change; original was syntactically correct.

I think if we can make it a little bit cleaner, that's good.

**Review by dgrove-oss — COMMENTED** [review]
Summary

Adds an explicit device parameter through spyre_empty_with_layout (Python and C++) so TP shard placement and Inductor-generated allocations select the correct Spyre device instead of relying on the ambient current device; also fixes a DeviceGuard/uncheckedSetDevice ordering bug in the allocator and an HBM-pool aliasing bug that could pool a fallback kernel's input storage.

Review

BLOCKERs

(none)

SUGGESTIONs

- Cross-device D2D shortcut in spyre_to doesn't check the destination device (torch_spyre/_monkey_patch.py:220-228). The else branch's no-op shortcut (if not copy and current_layout == device_layout: return self) only compares layouts, never target_device vs self.device. This PR adds the ability to pass an explicit device= into this exact code path but doesn't add a device check here — so tensor_on_spyre0.to("spyre:1", device_layout=same_layout) will silently return self, leaving the tensor on the wrong device instead of doing the D2D copy. This is precisely the new TP-placement scenario the PR targets (shards moving to a specific rank's device), so it seems worth closing. Suggest adding and current_device == target_device (or equivalent) to the shortcut condition, plus a regression test exercising same-layout-different-device .to().
- No test exercises .to(<device with explicit different index>, device_layout=...) at all — all existing and newly-added device_layout= call sites in tests/ use either no device or a bare "spyre" (ambient/ index-0 implied). Given the PR's stated purpose (per-rank TP placement), a test that places a shard on a non-current, non-zero device index via Tensor.to(..., device_layout=...) (as opposed to only via the new dma_*_to_spyre helpers, which are tested) would directly cover the spyre_to device-normalization changes and would have caught the gap above.

QUESTIONs

- dma_moe_expert_weight_to_spyre and dma_moe_per_expert_scale_to_spyre gained a device kwarg and call _normalize_spyre_device, but no new test exercises them with an explicit device (only dma_linear_weight_to_spyre/dma_embedding_weight_to_spyre get device-parametrized tests in test_model_utils.py). Intentional, given MoE TP sharding is presumably still coming through hf-adapters, or worth adding now for symmetry?

Verdict

COMMENT — CI is fully green, license/DCO/lint conventions are followed, the C++ DeviceGuard/uncheckedSetDevice fix and the HBM-pool aliasing fix are correct and well-tested. The one substantive concern is the untouched cross-device same-layout shortcut in spyre_to, which looks like a latent bug directly in the new feature's path — worth a decision from the author on whether to fix in this PR or file as a fast-follow before wider TP rollout.

**Comment by ani300** [comment]
@thoangtrvn @dgrove-oss I applied the current review round in `d106a3d4`.

Decisions made:

- Removed all three newly introduced public wrappers (`dma_tensor_to_spyre`, `dma_linear_weight_to_spyre`, and `dma_embedding_weight_to_spyre`). The existing private default/linear/embedding helpers now normalize and accept `device=` directly; hf-adapters PR 522 was updated to consume them.
- Kept the two pre-existing public MoE helpers and added explicit non-current-device coverage for both, alongside default, linear, and embedding layouts.
- Restored the prior import ordering/style and `_transfer_module` lambda formatting to avoid unrelated churn.
- Tightened the `Tensor.to(..., device_layout=...)` no-op condition to require the same device as well as the same layout.
- Fixed the D2D branch to return the allocated destination after the mutating copy operation.
- Added a real two-device same-layout regression that checks placement and validates data with a device-0 → device-1 → device-0 round trip.

Targeted result: 162 passed, 1 skipped, 6 subtests passed; Ruff and mypy hooks pass.

**Review by thoangtrvn — APPROVED** [review]
LGTM. Thanks @ani300 !

**Review by dgrove-oss — APPROVED** [review]
Summary

Adds an optional device parameter throughout the DMA-helper/allocation stack (spyre_empty_with_layout, model_utils.py helpers, Tensor.to(..., device_layout=...)) so TP shard loading can place weights directly on a target Spyre device with the correct layout, fixes a device-selection bug in the C++ allocator (DeviceGuard was installed too late, uncheckedSetDevice was a no-op), and fixes an HBM-pool-planning bug where an aliased buffer could steal the allocation backing a fallback kernel's input.

Review

I traced the three fixes through the actual code (not just the PR description):
- spyre_empty_with_layout in spyre_mem.cpp now takes device_opt, validates it's a PrivateUse1 device, and installs the DeviceGuard before constructing SpyreStorageImpl — matches the described bug (allocator reads guarded current device during storage construction).
- SpyreGuardImpl::uncheckedSetDevice now actually sets tls_idx, satisfying c10::DeviceGuard's contract to restore the prior device on scope exit.
- hbm_pool_planning.py's new fallback_read_alloc_ids set (keyed by id(layout.allocation), not by name) correctly closes the aliasing gap — this matches the new regression test test_alias_of_fallback_input_is_not_pool_eligible, which builds exactly the for_each_tile-style alias scenario described.
- spyre_to in _monkey_patch.py: the D2D no-copy shortcut now also requires self.device == dst.device, and when a copy is needed the D2D path correctly returns dst instead of the op's return value (a related but separate correctness fix — worth confirming this was intentional and not silently piggybacked, since torch.ops.spyre.copy_from_d2d's return value is discarded in the new code).
- Device string round-trips through torch.device(f"{device}") correctly (verified for indexed devices).
- All CI passed, DCO passed (5/5 commits signed off), stub (_C.pyi) signature matches the pybind11 binding exactly.

BLOCKERs

None found.

SUGGESTIONs

- [ ] In spyre_to's D2D branch, torch.ops.spyre.copy_from_d2d(...) result is now discarded and dst is returned directly instead. Confirm this is equivalent (i.e. the op mutates dst in place and its return value was always just dst) — worth a one-line comment since it's a subtle behavior change bundled into a device-awareness PR. (torch_spyre/_monkey_patch.py)

QUESTIONs

- [ ] test_dma_helpers_accept_an_explicit_device mocks copy_tensor and skips unless 2+ Spyre devices are present — is there any single-device CI lane that exercises the real (non-mocked) cross-device copy path, or is that only validated by the Gemma TP smoke test mentioned in the PR description?

Verdict

APPROVE — the three described bugs are real, each fix is narrowly scoped and matches its regression test, and CI is fully green with proper DCO signoff. The one suggestion above is worth a quick confirmation from the author but isn't blocking.
